misc/matrix_distribution.md

```diff
@@ -0,0 +1,92 @@
+
+The following constrains holds:
+- Each tile is stored in the memory of a single rank.
+- Each block of the 2D block cyclic distribution is a tile.
+
+## Naming conventions
+
+Strong type 2D indices and sizes are used to reduce the possibility of index mixing errors.
+The following naming convention is used for the types of indices and sizes:
+- `GlobalElementIndex`: index of an element in a global matrix.
+- `GlobalTileIndex`: index of a tile in a global matrix.
+- `LocalTileIndex`: index of a tile in a local matrix.
+- `TileElementIndex`: index of ain element in a tile.
+- `GlobalElementSize`: the size of a global matrix in elements.
+- `LocalElementSize`: the size of a local matrix in elements.
+- `GlobalTileSize`: the size of a global matrix in tiles.
+- `LocalTileSize`: the size of a local matrix in tiles.
+- `TileElementSize`: the size of a tile in elements.
+
+Warning: 1D indices and sizes have all the same type.
+
+## Index manipulation
+
+Utilities exist to tranform global indices to local indices and viceversa.
+The "Next" versions of conversions from global to local indices return the corresponding local index of the global index, if the global index is the given process,
+otherwise they return the local index corresponding to the next global index in the given process.
+
+The following table shows how a component of GlobalElementIndex tranforms for `block_size = 3`, `grid_size = 2`, `src_rank_index = 2`.
+
+| GlobalElementIndex        |  0 |  1 |  2 |  3 |  4 |  5 |  6 |  7 |  8 |  9 | 10 | 11 | 12 | 13 | 14 | 15 |
+|---------------------------|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
+| GlobalTileIndex           |  0 |  0 |  0 |  1 |  1 |  1 |  2 |  2 |  2 |  3 |  3 |  3 |  4 |  4 |  4 |  5 |
+| CommunicatorRank          |  1 |  1 |  1 |  2 |  2 |  2 |  0 |  0 |  0 |  1 |  1 |  1 |  2 |  2 |  2 |  0 |
+| LocalElementIndex Rank 0  |    |    |    |    |    |    |  0 |  1 |  2 |    |    |    |    |    |    |  3 |
+| LocalElementIndex Rank 1  |  0 |  1 |  2 |    |    |    |    |    |    |  3 |  4 |  5 |    |    |    |    |
+| LocalElementIndex Rank 2  |    |    |    |  0 |  1 |  2 |    |    |    |    |    |    |  3 |  4 |  5 |    |
+| LocalTileIndex Rank 0     |    |    |    |    |    |    |  0 |  0 |  0 |    |    |    |    |    |    |  1 |
+| LocalTileIndex Rank 1     |  0 |  0 |  0 |    |    |    |    |    |    |  1 |  1 |  1 |    |    |    |    |
+| LocalTileIndex Rank 2     |    |    |    |  0 |  0 |  0 |    |    |    |    |    |    |  1 |  1 |  1 |    |
+| NextLocalTileIndex Rank 0 |  0 |  0 |  0 |  0 |  0 |  0 |  0 |  0 |  0 |  1 |  1 |  1 |  1 |  1 |  1 |  1 |
+| NextLocalTileIndex Rank 1 |  0 |  0 |  0 |  1 |  1 |  1 |  1 |  1 |  1 |  1 |  1 |  1 |  2 |  2 |  2 |  2 |
+| NextLocalTileIndex Rank 2 |  0 |  0 |  0 |  0 |  0 |  0 |  1 |  1 |  1 |  1 |  1 |  1 |  1 |  1 |  1 |  2 |
+| TileElementIndex          |  0 |  1 |  2 |  0 |  1 |  2 |  0 |  1 |  2 |  0 |  1 |  2 |  0 |  1 |  2 |  0 |
```

Comment:
Is there a reason why you specified values for the `TileElementIndex` without splitting it as viewed by each rank as you did for other examples?

Reply:
All the ranks gets the same answer.